File: merge_sort.rb
# frozen_string_literal: true

def merge_sort(array)
  n = array.length
  return array if n <= 1

  mid = n / 2
  left_array = array[0..mid - 1]
  right_array = array[mid..n - 1]

  sorted_left = merge_sort(left_array)
  sorted_right = merge_sort(right_array)
  merge(array, sorted_left, sorted_right)

  array
end

def merge(array, sorted_left, sorted_right)
  left_length = sorted_left.length
  right_length = sorted_right.length

  array_index = 0
  left_index = 0
  right_index = 0

  while left_index < left_length && right_index < right_length
    if sorted_left[left_index] < sorted_right[right_index]
      array[array_index] = sorted_left[left_index]
      left_index += 1
    else
      array[array_index] = sorted_right[right_index]
      right_index += 1
    end
    array_index += 1
  end

  while left_index < left_length
    array[array_index] = sorted_left[left_index]
    left_index += 1
    array_index += 1
  end

  while right_index < right_length
    array[array_index] = sorted_right[right_index]
    right_index += 1
    array_index += 1
  end

  array
end

array = [105, 79, 100, 110]

p merge_sort(array)

File: merge_sort.js
function mergeSort(array) { 
    const n = array.length;
    if (n <= 1 ) { 
        return array;
    }

    const mid = Math.round(n/2);
    const leftArray = array.slice(0, mid)
    const rightArray = array.slice(mid);

    const sortedLeft = mergeSort(leftArray);
    const sortedRight = mergeSort(rightArray);

    const sortedArray = merge(array, sortedLeft, sortedRight);

    return sortedArray;
}

function merge(array, sortedLeft, sortedRight) { 
    const leftLength = sortedLeft.length;
    const rightLength = sortedRight.length

    let arrayIndex = 0;
    let leftIndex = 0;
    let rightIndex = 0;

    while ( leftIndex < leftLength && rightIndex < rightLength ) {
        if (sortedLeft[leftIndex] < sortedRight[rightIndex]) {
            array[arrayIndex] = sortedLeft[leftIndex];
            leftIndex += 1;
        } else {
            array[arrayIndex] = sortedRight[rightIndex];
            rightIndex += 1;
        }
        arrayIndex += 1 // may have to nest under both if and else
    }

    while ( leftIndex < leftLength) {
        array[arrayIndex] = sortedLeft[leftIndex];
        leftIndex += 1;
        arrayIndex += 1;
    }

    while ( rightIndex < rightLength ) { 
        array[arrayIndex] = sortedRight[rightIndex];
        rightIndex += 1
        arrayIndex += 1
    }

    return array
}

/*
mergeSort([]) → []
mergeSort([73]) → [73]
mergeSort([1, 2, 3, 4, 5]) → [1, 2, 3, 4, 5]
mergeSort([3, 2, 1, 13, 8, 5, 0, 1]) → [0, 1, 1, 2, 3, 5, 8, 13]
mergeSort([105, 79, 100, 110]) → [79, 100, 105, 110]
*/

console.log(mergeSort([105, 79, 100, 110]))
